url146c819f25_jlink_Parse4SJM_tgz-pJ8-sjm_examples_book_robot_ShowRobotRefactoredJ8/src/sjm/examples/book/regular/RegularParser.java

```diff
@@ -0,0 +1,214 @@
+	 *
+	 *    nextFactor = factor;
+	 *
+	 * This parser has an assembler that will pop two
+	 * parsers and push a Sequence of them. 
+	 */
+	protected Parser nextFactor() {
+		Parser p = factor();
+		p.setAssembler(new AndAssembler());
+		return p;
+	}
+
+	/*
+	 * Returns a parser that for the grammar rule:
+	 *
+	 *    orTerm = '|' term;
+	 *
+	 * This parser has an assembler that will pop two
+	 * parsers and push an Alternation of them. 
+	 */
+	protected Parser orTerm() {
+		Sequence s = new Sequence();
+		s.add(new SpecificChar('|').discard());
+		s.add(term());
+		s.setAssembler(new OrAssembler());
+		return s;
+	}
+
+	/*
+	 * Returns a parser that for the grammar rule:
+	 *
+	 *     phrase = letterOrDigit | '(' expression ')';
+	 */
+	protected Parser phrase() {
+		Alternation a = new Alternation();
+		a.add(letterOrDigit());
+
+		Sequence s = new Sequence();
+		s.add(new SpecificChar('(').discard());
+		s.add(expression());
+		s.add(new SpecificChar(')').discard());
+
+		a.add(s);
+		return a;
+	}
+
+	/*
+	 * Returns a parser that for the grammar rule:
+	 *
+	 *    phraseStar = phrase '*'; 
+	 *
+	 * This parser has an assembler that will pop a
+	 * parser and push a Repetition of it.
+	 */
+	protected Parser phraseStar() {
+		Sequence s = new Sequence();
+		s.add(phrase());
+		s.add(new SpecificChar('*').discard());
+		s.setAssembler(new StarAssembler());
+		return s;
+	}
+
+	/**
+	 * Returns a parser that will recognize a regular
+	 * expression. 
+	 * 
+	 * @return a parser that will recognize a regular
+	 *         expression
+	 */
+	public static Parser start() {
+		return new RegularParser().expression();
+	}
+
+	/*
+	 * Returns a parser that for the grammar rule:
+	 *
+	 *    term = factor nextFactor*; 
+	 */
+	protected Parser term() {
+		Sequence term = new Sequence();
+		term.add(factor());
+		term.add(new Repetition(nextFactor()));
+		return term;
+	}
+
+	/**
+	 * Return a parser that will match a <code>
+	 * CharacterAssembly</code>, according to the value of a 
+	 * regular expression given in a string.
+	 *
+	 * For example, given the string <code>a*</code>, this 
+	 * method will return a parser which will match any element
+	 * of the set <code>{"", "a", "aa", "aaa", ...}</code>.
+	 *
+	 * @return a parser that will match a <code>
+	 *         CharacterAssembly</code>, according to the value
+	 *         of a regular expression in the given string
+	 *
+	 * @param   String   the string to evaluate
+	 *
+	 * @exception RegularExpressionException if this parser
+	 *            does not recognize the given string as a 
+	 *            valid expression
+	 */
+	public static Parser value(String s) throws RegularExpressionException {
+
+		CharacterAssembly c = new CharacterAssembly(s);
+		Assembly a = start().completeMatch(c);
+		if (a == null) {
+			throw new RegularExpressionException("Improperly formed regular expression");
+		}
+		Parser p;
+		try {
+			p = (Parser) a.pop();
+		} catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)